packages/suite/src/components/suite/Preloader/index.tsx

```diff
@@ -94,8 +95,8 @@ const getSuiteApplicationState = ({
     if (authConfirmation?.type === 'auth-confirm') return DiscoveryLoader;
 };
 
-const getModalApplication = (route: AppState['router']['route']) => {
-    if (!route) return;
+/** For given route that is a modal returns corresponding View. */
+const getModalApplication = (route: ModalRoute): ((...args: any[]) => JSX.Element | null) => {
```

Comment:
```suggestion
const getModalApplication = (route: ModalRoute) => {
```
unnecessary return type

Reply:
wouldn't be easier to have `router.route?.isModal` check inside this function and leave the rest as it was before?
introducing new type declared in different file and used only here looks kinda unnecessary

i mean something like previous line 98 `if (!route || !route.isModal) return;`

Reply:
As explained in the [commit message ](https://github.com/trezor/trezor-suite/pull/3590/commits/517a147e0c6d1c393e023cdf06154be9bd4359cd) the motivation behind this is to make this function typesafe in a sense that TS compiler will barf if all possible cases aren't handled in the `switch`. This should help us in two ways:
1) Future devs won't spent time (as I did) figuring out why their newly added modal won't open with `goto()` when they didn't added a case to this switch.
2) It will prevent cases that can never happen and thus will keep the `switch` clean and minimal. This is exactly what happened with the `log` case. .)

For this to work we must tell the compiler that this function must always return a modal it will then ensure that the switch is exhaustive.

OTOH I agree that the return type *is* ugly. I'm open to any suggestion on making this nicer. `((...args: any[]) => JSX.Element | null)` is the only intersection of all our modals I could come up with. Maybe we can do better?

I also agree that definning `ModalRoute` could be done here. Either separately
```js
type ModalRoute = Extract<typeof routes[number], { isModal: true }>;
const getModalApplication = (route: ModalRoute): ...
```
or inline
```js
const getModalApplication = (route: Extract<typeof routes[number], { isModal: true }>): ...
```
I don't have a strong opinion on neither of these two.

Reply:
if you use suggested solution you will get exactly the same results without need to declare additional strict type.
[try it out here](https://www.typescriptlang.org/play?ssl=68&ssc=30&pln=67&pc=12#code/PTAEFp1BXAXBLANgZ1M2BDATrAULgUwA8AHAex1FgE8SDQBBLLDagUUQIFsCA7WADxMW1ACq16xWHwAmqLAQwyyvRNRi8A1rzIB3XgG0AugD5QAXkbNW4ugd7QuAIwJYjAbnzFylGnVAAwiroAOrwsAAWAPIkCCoYiABi8ASIcgK4oFmBwbCgUrKoAN6gBpoE1ABcaLBY8LwA5kbVGLzqAL4ANJnZyaly+UTSvAPodY2gAD6gDs6uU2jUzmSIuGaWRT1ZBgDSFaD1oOXUZABmObzozaCb2XelfWl76oePctdBl7C7FUYGb88jINhnItvc7tMxvUGmDwVlprMXFhYXDpiUyhVqlDGtdWh0UeDpk4yCtFLwCfcAPwXdA-ah-AG-Cl3arQEYEU71AgyTx3dqefn4EAQKBwJCoWRCsCQUBYMhwebobB4XAAY1ysvl0lQlgMYJKGBIJGqAHIdLBwFwyDIEuAAIwm0BdfWgQ3G0BmsgWq02xDgABMjuddwNRtN5st1ttAGYg90Q66wx6nHUZA0CCbOgdkABZKOIaq1aD0YPZUPuk3QGQEABumezed9hawxad8bLiYrNdcyHgKnr8Fz+ebredQIwqHVX08uD89AASlqCM8dUcKmcrCIONw+II5xu5QrkGZpibB42Eo7TyRsBguMgTZ5vBQ8nPQIuFRZN6xtzx+AJPlCcJoliPteASN5kAEfdzkPbUsw-aQVxME8NGrTleG5GdhRlOD5klWcJFAAAFOUSFXW5slVVpVVSDAnE4apiVJVpeWyU4SWqRFXDYrInGwLFamhAUpVlRRVTyQcZi9V0awwJB6M4UAIlcAgs04WBUBIBQQSoCIMFfCIpKkjBQHVLhyEw-g1Q1C9EDtL8AAptLIcjqlI1zkAASgsMxKKyHToCwXgPUSKIokdCczNydxQGFCg0DIHhQAAKQAZQADQAOl-XdcHafAp3QUA7P9JyXLckiyO83ybjBQLgo9AAhBh50iycYrisAEuQJL6HS7LcusgrCts-No3K6r3OqnzzD8+qCFgIKQocRBEFi+KsES5KBpyzg-zwEavFIF9oq+UB01gOyGCNRB4GouIQssRy8OqRCCFm+a7ngc5HIAQjwqZpgBpcsvPfMfIa3heLQXRwlVCJQBe0G3R8-z7mo5B6BNFN4DTDNKmZbIoZK-M7Rhu5MexqtaxNQm4XuEnSop7JhSpj1uywXt+3puFhTuJnxpnBnhR0UB0IwaBEDwPl8vcIA) and comment one of the switch cases.

Regarding function return type, typescript will figure it out by itself and returned values will be strict (again, see example above)
returned components will require assigned props, like cancelable

i guess by resoving (2) we also resolve (1) as well :)

Reply:
> turns out the real key to enable strict switch was removing default case :)

Yes, but only when [`noImplicitReturns`](https://www.typescriptlang.org/tsconfig#noImplicitReturns) TS option is set to `true`. In Suite, we currently leave it to the default which is `false`. Enabling this lint raises some 65 errors in the Suite package. I'm up for doing that but in another PR. [Here](https://github.com/trezor/trezor-suite/issues/3630)'s an issue for that. Question is: do we want to keep the commit in question in this PR or defer it to the issue as well?

Reply:
i've just checked on your branch that `ModalRoute` type does not work for exactly same reason.
the only reason why it throws error here is because you've defined return type of the `getModalApplication` function.
problem here is this return type is not precise, it returns `JSX.Element` with `any` props, therefore `<ApplicationModal` (line 177) will accept `any` props (example: set `cancelable="foo"` - no error, but it should be)

Reply:
> the only reason why it throws error here is because you've defined return type of the getModalApplication function.

Yup, exactly.

> problem here is this return type is not precise

Aha! Well I could either edit the signature to
```typescript
const getModalApplication = (route: ModalRoute): ((a: InjectedModalApplicationProps) => JSX.Element | null) => {...}
```

or remove this commit from this PR and defer this refactor to later when `noImplicitReturns` is enabled and we don't have to hack around with explicit return types. I'm up for the later. What you think?

Reply:
noImplicitReturns PR would the right place for that i think

Reply:
Ok, removed the commit. 

I force-pushed here, as it seemed over-complicated to have a revert fixup commit. Hope that's ok in this case. 👉 👈